### PR TITLE
LinearAlgebra.givens with unitful arguments

### PR DESCRIPTION
**Pull request by KlausC**
Solves JuliaLang/LinearAlgebra.jl#738

### CODE REVIEW

**Comment by StefanKarpinski** [comment]
Thanks! Would be great to add the example from the issue as a test.

**Comment by KlausC** [comment]
The example is from external package `Unitful`. Is there a type `T` in `Base` with `one(T) != oneunit(T)`?

**Comment by StefanKarpinski** [comment]
Right. Sometimes we'll mock up very minimal types for this kind of testing. Not sure if we have one already...

**Comment by KlausC** [comment]
> Right. Sometimes we'll mock up very minimal types for this kind of testing. Not sure if we have one already...

I did not find one. Will make a minimal type similar to `Unitful.Quantity` in module `TestGivens`.

**Comment by StefanKarpinski** [comment]
This seems pretty simple, but it would be good if someone who actually knows what a Givens rotation is took a look. @andreasnoack, could you review?

File: stdlib/LinearAlgebra/src/givens.jl [hunk 1]
```diff
@@ -248,6 +249,18 @@ function givensAlgorithm(f::Complex{T}, g::Complex{T}) where T<:AbstractFloat
...
+# enable for unitful quantities
+function givensAlgorithm(f::T, g::T) where T
+    fs = f / oneunit(T)
+    gs = g / oneunit(T)
+    typeof(fs) === T && typeof(gs) === T &&
+    !isa(fs, Union{AbstractFloat,Complex{<:AbstractFloat}}) &&
+    throw(MethodError(givensAlgorithm, (fs, gs)))
+
+    c, s, r = givensAlgorithm(fs, gs)
+    c, s, r * oneunit(T)
```
Comment (dkarrasch) [inline stdlib/LinearAlgebra/src/givens.jl:261]: Nitpicking, but the above methods do have a `return` statement.
```suggestion
    return c, s, r * oneunit(T)
```

**Review by dkarrasch — COMMENTED** [review]
LGTM, except for one minor style comment.

**Review by andreasnoack — APPROVED** [review]

**Review by dkarrasch — APPROVED** [review]

**Review by dkarrasch — CHANGES_REQUESTED** [review]
There is some nasty method ambiguity error, so this cannot be merged as is, unfortunately. It's a strange error, because it seems to relate to multiplication methods defined in the tests, not in the code. Would it be possible to reformulate the MockUnitful tests in terms of Furlongs?

**Comment by KlausC** [comment]
> Would it be possible to reformulate the MockUnitful tests in terms of Furlongs?

Unfortunately not, because `Furlong` does not have the property that `f / oneunit(f) isa Union{AbstractFloat,Complex{<:AbstractFloat}}` for `f::Furlong`, which is needed to call `givensAlgorithm`.

The fix assumes, that `f / oneunit(f)` strips off the unit from the type of `f`; there seems no other way inside `Base` to accomplish that :-1: . (I could imagine something like `stripunit` as a companion of `oneunit` to do that).

The detected ambiguities are an insane interaction of `Furlong` and `MockUnitful`. I change the definition of `MockUnitful` to get rid of it. If they are detected or not seems to be a timing problem, when test-ambiguities is still working while test-LinearAlgebra is executed.

**Review by dkarrasch — APPROVED** [review]

**Comment by stevengj** [comment]
> Unfortunately not, because `Furlong` does not have the property that `f / oneunit(f)` isa `Union{AbstractFloat,Complex{<:AbstractFloat}}` for `f::Furlong`, which is needed to call `givensAlgorithm`.

Arguably this was a bug in the `Furlong` test type, which should have been fixed rather than defining yet another unitful test type.   It's fixed in #39852.

**Pull request merged**